### PR TITLE
Support VS 2017

### LINKED ISSUE #38584: Add support for the Visual Studio 2017 linker
With Visual Studio 2017, Microsoft has changed the folder structure it uses for everything. `link.exe` is now located at `C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.10.24728\bin\HostX64\x64\link.exe`. As you can see, the folder is dependent on the actual edition of Visual Studio as well.

Currently, Rust does not seem to recognize this path correctly. I don’t know exactly what Rust needs to figure out the paths, but if you need some information (e.g. registry values, contents of `vcvarsall.bat` or something), just tell me what you need and I’ll try my best to provide you with it.

### PR DESCRIPTION
**Pull request by brson**
Fixes #38584 

This replaces all the MSVC linker logic with that from the 'gcc' crate. The code looks the same, but there could be regressions.

I've only tested this with x86_64.

r? @alexcrichton 
cc @vadimcn @retep998

### CODE REVIEW

**Comment by brson** [comment]
I'm seeing issues in cmake that i want to investigate still.

**Comment by brson** [comment]
Ok, I think this patch is sufficient to fix rustc at least. Both gcc-rs and cmake still need to be updated to be able to find msbuild correctly. I'll file separate issues on those.

**Comment by brson** [comment]
The [rustup patch](https://github.com/rust-lang-nursery/rustup.rs/pull/1136) failed on i686.

**Comment by ishitatsuyuki** [comment]
Rollup failure.

PS: @frewsxcv please be careful when rolling up non-rollup priority PRs, inspect the PR comments before adding it to list.

**Comment by brson** [comment]
I'm not sure how the appveyor images are set up, but I suspect that after this lands we will be building Rust releases with VS 2017 instead of VS 2015.

**Comment by est31** [comment]
The error from the rollup is:

```
error: linking with `link.exe` failed: exit code: 1171
  |
  = note: "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\bin\\amd64_x86\\link.exe" "/NOLOGO" "/NXCOMPAT" "/LARGEADDRESSAWARE" "/SAFESEH" "/LIBPATH:C:\\projects\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "C:\\projects\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\deps\\std-ceb42dcce624041e.0.o" "/OUT:C:\\projects\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\deps\\std-ceb42dcce624041e.dll" "/DEF:C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.RkAu4DLaw1LX\\lib.def" "C:\\projects\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\deps\\std-ceb42dcce624041e.crate.metadata.o" "/OPT:REF,ICF" "/DEBUG" "/LIBPATH:C:\\projects\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\deps" "/LIBPATH:C:\\projects\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\release\\deps" "/LIBPATH:C:\\projects\\rust\\build\\i686-pc-windows-msvc\\native\\compiler-rt\\." "/LIBPATH:C:\\projects\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "advapi32.lib" "ws2_32.lib" "userenv.lib" "shell32.lib" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.RkAu4DLaw1LX\\librand-cbaabb050663bb1e.rlib" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.RkAu4DLaw1LX\\libpanic_unwind-7a8f267e1f472275.rlib" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.RkAu4DLaw1LX\\liblibc-913fac93272423bb.rlib" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.RkAu4DLaw1LX\\libcollections-0e863adcb979d99b.rlib" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.RkAu4DLaw1LX\\libunwind-f8f46edb99891d80.rlib" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.RkAu4DLaw1LX\\libstd_unicode-72685d3c05b2de5b.rlib" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.RkAu4DLaw1LX\\liballoc-2fe699c323a9fcd5.rlib" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.RkAu4DLaw1LX\\liballoc_system-1a9ffce435c6813e.rlib" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.RkAu4DLaw1LX\\libcore-6796d5cf3cb4eed3.rlib" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.RkAu4DLaw1LX\\libcompiler_builtins-13752c636400e0ca.rlib" "libcmt.lib" "/DLL" "/IMPLIB:C:\\projects\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\deps\\std-ceb42dcce624041e.dll.lib"
  = note: LINK : fatal error LNK1171: unable to load mspdbcore.dll (error code: 126)      
```

The version number 15.0 corresponds to 2015, not 2017. And the linker is the only thing from VS we are running, right?

**Comment by brson** [comment]
Huh. That's not the error I expected :-/

**Comment by retep998** [comment]
>The version number 15.0 corresponds to 2015, not 2017. And the linker is the only thing from VS we are running, right?

No, 15 corresponds to 2017 and 14 corresponds to 2015.

>note: LINK : fatal error LNK1171: unable to load mspdbcore.dll (error code: 126) 

This error is typically caused when you're running a cross linker but you forgot to add the host toolchain to `PATH` (because the cross linker folder depends on some DLLs from the host toolchain folder).

File: src/librustc_trans/back/link.rs [hunk 1]
```diff
@@ -150,14 +149,35 @@ pub fn get_linker(sess: &Session) -> (String, Command, Option<PathBuf>) {
...
+#[cfg(windows)]
+pub fn msvc_link_exe_cmd(sess: &Session) -> (Command, Option<PathBuf>) {
+    use gcc::windows_registry;
+
+    let target = &sess.opts.target_triple;
+    let tool = windows_registry::find_tool(target, "link.exe");
+
+    if let Some(tool) = tool {
+        let path = tool.path().parent().map(|p| p.to_owned());
+        (tool.to_command(), path)
```
Comment (retep998) [inline src/librustc_trans/back/link.rs]: The issue is right here. `path` is *not* supposed to be the directory with the linker, but rather the directory of the *host toolchain*. You'll have to teach `gcc` how to provide that information so you can pull it out here.

**Comment by est31** [comment]
> No, 15 corresponds to 2017 and 14 corresponds to 2015.

[Right](https://en.wikipedia.org/wiki/Microsoft_Visual_Studio#History). I have mistyped. The version number of the pasted text is 14.0 though, so it still corresponds to 2015.

File: src/librustc_trans/back/link.rs (hunk 1 above)
Comment (brson) [inline src/librustc_trans/back/link.rs]: Thanks for the review @retep998

**Comment by brson** [comment]
@bors r+

I want to see what bors thinks again.

This is currently using a git replacement of the gcc crate.

**Comment by brson** [comment]
@bors r+

I changed the travis config to fail so this won't land with the crate replacement.

**Comment by brson** [comment]
Still just trying to see what bors thinks.

**Comment by retep998** [comment]
It appears some travis builders finished so quickly that the appveyor bots didn't have time to finish.

**Comment by kennytm** [comment]
https://travis-ci.org/rust-lang/rust/jobs/238282859

```sh
$ sleep 300000

No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.

Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

What is this `sleep 300000` thing doing here?

**Comment by retep998** [comment]
@kennytm It's preventing the CI build from succeeding because brson wants to know how the appveyor CI likes the changes without actually landing the changes yet.

**Comment by kennytm** [comment]
@retep998 yes I understand, that's the comment was deleted afterwards.

**Comment by brson** [comment]
😞

**Comment by brson** [comment]
@bors r+

Again just trying to get a little love from bors.

**Comment by Mark-Simulacrum** [comment]
AppVeyor passed; I'm canceling this. https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3454

@bors retry 
@bors r-

**Comment by brson** [comment]
This is ready to go now.

r? @alexcrichton

**Comment by Mark-Simulacrum** [comment]
@bors rollup

We know that this passes AppVeyor, and Travis PR passed (and I think we've passed regular travis before).

**Comment by mosabbirkhan** [comment]
Hi! I’ve signed up for a spam filtering service called BitBounce and I added you to my “whitelist”. That means you can email me and your emails will come through.

For people who aren’t on my whilelist, I’m charging $0.05 in Credo for their email to get to my inbox. A cleaner inbox and more money? Win! You should check it out!

Sign up to get paid for emails here: https://bitbounce.com?ref=bitbounce


BitBounce is a product by:

Turing Technology, Inc.
BitBounce.com
Redwood City, CA

Unsubscribe from Dora Channe5's whitelist notification emails here: https://bitbounce.com/unsubscribe_from_whitelist_email_notifications/1683488778

**Pull request merged**